Content.Server/Roles/RoleTimerSystem.cs

```diff
@@ -0,0 +1,306 @@
+using Content.Server.Afk.Events;
+using Content.Server.GameTicking;
+using Content.Server.Players;
+using Content.Shared.CCVar;
+using Content.Shared.GameTicking;
+using Content.Shared.Roles;
+using Robust.Server.GameObjects;
+using Robust.Server.Player;
+using Robust.Shared.Enums;
+using Robust.Shared.Network;
+using Robust.Shared.Player;
+using Robust.Shared.Timing;
+using Robust.Shared.Utility;
+
+namespace Content.Server.Roles;
+
+/// <summary>
+/// This handles issuing saves of role / overall times to the DB during the regular course of play.
+/// <see cref="RoleTimerManager"/> handles the actual data.
+/// </summary>
+public sealed class RoleTimerSystem : SharedRoleTimerSystem
+{
+    [Dependency] private readonly IAfkManager _afk = default!;
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] private readonly IPlayerManager _playerManager = default!;
+    [Dependency] private readonly GameTicker _ticker = default!;
+    [Dependency] private readonly RoleTimerManager _roleTimers = default!;
+
+    /// <summary>
+    /// Autosave regularly in case of server crash.
+    /// </summary>
+    private float _autosaveDelay = 900;
+
+    private float _autoSaveAccumulator;
+
+    /// <summary>
+    /// If someone just joined track the last time we set their times so the autosave doesn't round up.
+    /// </summary>
+    private readonly Dictionary<IPlayerSession, TimeSpan> _lastSetTime = new();
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        ConfigManager.OnValueChanged(CCVars.GameRoleTimersSaveFrequency, SetAutosaveDelay, true);
+        _playerManager.PlayerStatusChanged += OnPlayerStatusChanged;
+        SubscribeLocalEvent<RoundRestartCleanupEvent>(OnRoundEnd);
+        SubscribeLocalEvent<PlayerAttachedEvent>(OnPlayerAttached);
+        SubscribeLocalEvent<PlayerDetachedEvent>(OnPlayerDetached);
+        SubscribeLocalEvent<RoleAddedEvent>(OnRoleAdd);
+        SubscribeLocalEvent<RoleRemovedEvent>(OnRoleRemove);
+        SubscribeLocalEvent<AFKEvent>(OnAFK);
+        SubscribeLocalEvent<UnAFKEvent>(OnUnAFK);
+    }
+
+    public IEnumerable<string> GetTimedRoles(Mind.Mind mind)
+    {
+        foreach (var role in mind.AllRoles)
+        {
+            if (role is not IRoleTimer timer) continue;
+            yield return ProtoManager.Index<RoleTimerPrototype>(timer.Timer).ID;
+        }
+    }
+
+    private List<string> GetTimedRoles(IPlayerSession session)
+    {
+        var roles = new List<string>();
+        var contentData = _playerManager.GetPlayerData(session.UserId).ContentData();
+
+        if (contentData?.Mind == null) return roles;
+
+        foreach (var mindRole in GetTimedRoles(contentData.Mind))
+        {
+            roles.Add(mindRole);
+        }
+
+        return roles;
+    }
+
+    private void OnRoleRemove(RoleRemovedEvent ev)
+    {
+        if (ev.Mind.Session == null) return;
+        Save(ev.Mind.Session, _timing.CurTime);
```

Comment:
Full database commit every time somebody changes their roles slightly seems excessive. This is going to cause a massive spike of DB commits on something like round restart.

Reply:
The roles are already removed from the mind's role list by the time this event is raised. This is partially the fault of the event not being named in past tense.

Either way, this code is incorrect. `Save()` won't find the role because it's already not in the list anymore, so some roles will fail to be accounted for.

Reply:
I would argue that role timer info should use **real time**, not simulation timing. This is mostly a player facing matter, so punishing players for any server timer disparities seems wrong.